File: 2025/07 July/07 Next Greater Element in Circular Array/main.cpp
// Link: https://www.geeksforgeeks.org/problems/next-greater-element/1

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution
{
public:
    vector<int> nextLargerElement(vector<int> &arr)
    {
        int n = arr.size();
        vector<int> res(n, -1);
        stack<int> stk;

        for (int i = 2 * n - 1; i >= 0; i--)
        {
            while (!stk.empty() && stk.top() <= arr[i % n])
            {
                stk.pop();
            }

            if (i < n && !stk.empty())
            {
                res[i] = stk.top();
            }

            stk.push(arr[i % n]);
        }

        return res;
    }
};

int main()
{
    Solution sol;
    vector<int> arr = {1, 3, 2, 4};
    vector<int> result = sol.nextLargerElement(arr);
    for (int i : result)
    {
        cout << i << " ";
    }
    cout << endl;

    arr = {0, 2, 3, 1, 1};
    result = sol.nextLargerElement(arr);
    for (int i : result)
    {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}

File: 2025/07 July/07 Next Greater Element in Circular Array/main.py
# Link: https://www.geeksforgeeks.org/problems/next-greater-element/1


class Solution:
    def nextLargerElement(self, arr):
        n = len(arr)
        res = [-1] * n
        stk = []

        for i in range(2 * n - 1, -1, -1):
            while stk and stk[-1] <= arr[i % n]:
                stk.pop()

            if i < n and stk:
                res[i] = stk[-1]

            stk.append(arr[i % n])

        return res


if __name__ == "__main__":
    sol = Solution()
    arr = [1, 3, 2, 4]
    result = sol.nextLargerElement(arr)
    print("Next Greater Elements:", result)

    arr = [0, 2, 3, 1, 1]
    result = sol.nextLargerElement(arr)
    print("Next Greater Elements:", result)
